### ISSUE TITLE
8. Real-Time Interface with PyBullet, ImGui, and OpenCV

### ISSUE THREAD

**Issue by ricardoamartinez** [ricardoamartinez/koopman-physics-neural-cellular-automata#9]
# Real-Time Interface with PyBullet, ImGui, and OpenCV

## Overview
Develop a **real-time interface** for visualizing and interacting with the agent during simulation. This combines OpenCV for frame rendering, ImGui for interactive controls, and PyBullet integration to provide live monitoring, parameter adjustment, and comprehensive visualization of the agent's internal states.

### Core Interface Architecture
Integrate OpenCV display with ImGui controls in the main loop:

```python
import cv2
import imgui
import imgui.integrations.opengl as gl_imgui
import OpenGL.GL as gl
import numpy as np

class RealTimeInterface:
    def __init__(self, window_width=1280, window_height=720):
        self.window_width = window_width
        self.window_height = window_height
        self.setup_imgui()
        self.setup_opencv()
        
        # Interface state variables
        self.training_enabled = True
        self.learning_rate = 1e-3
        self.show_debug_info = True
        self.visualization_mode = "rgb"
        
    def setup_imgui(self):
        """Initialize ImGui context and renderer"""
        imgui.create_context()
        self.impl = gl_imgui.FixedPipelineRenderer()
        
        # Configure ImGui style
        style = imgui.get_style()
        style.window_rounding = 5.0
        style.child_rounding = 5.0
        style.frame_rounding = 3.0
        
    def setup_opencv(self):
        """Setup OpenCV windows for visualization"""
        cv2.namedWindow("Simulation View", cv2.WINDOW_AUTOSIZE)
        cv2.namedWindow("Latent Visualization", cv2.WINDOW_AUTOSIZE)
```

### PyBullet Camera Rendering
Extract and display camera images from PyBullet:

```python
def render_pybullet_camera(view_matrix, proj_matrix, img_width=320, img_height=320):
    """Render camera view from PyBullet simulation"""
    _, _, rgb_img, depth_img, _ = p.getCameraImage(
        img_width, img_height,
        viewMatrix=view_matrix,
        projectionMatrix=proj_matrix
    )
    
    # Convert to OpenCV format
    rgb_array = np.array(rgb_img, dtype=np.uint8).reshape(img_height, img_width, 4)
    frame_bgr = cv2.cvtColor(rgb_array, cv2.COLOR_RGBA2BGR)
    
    # Add overlay information
    frame_with_overlay = add_simulation_overlay(frame_bgr, depth_img)
    
    return frame_with_overlay

def add_simulation_overlay(frame, depth_img):
    """Add informational overlay to simulation frame"""
    overlay_frame = frame.copy()
    
    # Add timestamp
    timestamp = time.strftime("%H:%M:%S")
    cv2.putText(overlay_frame, f"Time: {timestamp}", (10, 30), 
                cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0, 255, 0), 2)
    
    # Add depth visualization (corner inlay)
    depth_normalized = (depth_img * 255).astype(np.uint8)
    depth_colored = cv2.applyColorMap(depth_normalized, cv2.COLORMAP_JET)
    
    # Resize depth for corner overlay
    small_depth = cv2.resize(depth_colored, (80, 60))
    overlay_frame[10:70, 250:330] = small_depth
    
    # Add border around depth inlay
    cv2.rectangle(overlay_frame, (250, 10), (330, 70), (255, 255, 255), 2)
    cv2.putText(overlay_frame, "Depth", (255, 85), 
                cv2.FONT_HERSHEY_SIMPLEX, 0.4, (255, 255, 255), 1)
    
    return overlay_frame
```

### ImGui Control Panels
Create interactive control panels for real-time parameter adjustment:

```python
def render_control_panel(self, agent, optimizer, constraint_manager):
    """Render main control panel with training controls"""
    imgui.new_frame()
    
    # Main Control Window
    if imgui.begin("Agent Control Panel", True)[1]:
        # Training Controls
        imgui.text("Training Controls")
        imgui.separator()
        
        # Training toggle
        clicked, self.training_enabled = imgui.checkbox("Enable Training", self.training_enabled)
        
        # Learning rate slider
        changed, self.learning_rate = imgui.slider_float(
            "Learning Rate", self.learning_rate, 1e-5, 1e-1, "%.1e"
        )
        if changed and optimizer:
            for param_group in optimizer.param_groups:
                param_group['lr'] = self.learning_rate
        
        # Exploration noise
        changed, exploration_noise = imgui.slider_float(
            "Exploration Noise", getattr(agent, 'exploration_noise', 0.1), 0.0, 1.0
        )
        if changed:
            agent.exploration_noise = exploration_noise
        
        imgui.spacing()
        
        # Visualization Controls
        imgui.text("Visualization")
        imgui.separator()
        
        _, self.show_debug_info = imgui.checkbox("Show Debug Info", self.show_debug_info)
        
        # Visualization mode selection
        imgui.text("Camera Mode:")
        imgui.same_line()
        if imgui.radio_button("RGB", self.visualization_mode == "rgb"):
            self.visualization_mode = "rgb"
        imgui.same_line()
        if imgui.radio_button("Depth", self.visualization_mode == "depth"):
            self.visualization_mode = "depth"
        imgui.same_line()
        if imgui.radio_button("Segmentation", self.visualization_mode == "seg"):
            self.visualization_mode = "seg"
        
        imgui.spacing()
        
        # Constraint Controls
        if constraint_manager:
            imgui.text("Constraints")
            imgui.separator()
            
            for name, lambda_val in constraint_manager.lambdas.items():
                # Display current constraint values
                imgui.text(f"{name}: λ = {lambda_val.item():.4f}")
                
                # Allow manual adjustment
                changed, new_val = imgui.slider_float(
                    f"##lambda_{name}", lambda_val.item(), 0.0, 10.0
                )
                if changed:
                    constraint_manager.lambdas[name].data.fill_(new_val)
        
    imgui.end()

def render_performance_monitor(self, training_metrics):
    """Render performance monitoring window"""
    if imgui.begin("Performance Monitor", True)[1]:
        # Training metrics
        if 'loss_history' in training_metrics:
            losses = training_metrics['loss_history'][-100:]  # Last 100 values
            if losses:
                imgui.plot_lines("Training Loss", losses, scale_min=0.0)
        
        # FPS monitoring
        current_fps = training_metrics.get('fps', 0.0)
        imgui.text(f"FPS: {current_fps:.1f}")
        
        # Memory usage
        if torch.cuda.is_available():
            memory_used = torch.cuda.memory_allocated() / 1e9  # GB
            memory_total = torch.cuda.memory_reserved() / 1e9   # GB
            imgui.text(f"GPU Memory: {memory_used:.2f} / {memory_total:.2f} GB")
        
        # Model statistics
        if hasattr(training_metrics, 'gradient_norm'):
            imgui.text(f"Gradient Norm: {training_metrics['gradient_norm']:.6f}")
        
    imgui.end()

def render_latent_visualization(self, latent_state):
    """Render latent state visualization"""
    if imgui.begin("Latent State Visualization", True)[1]:
        if latent_state is not None:
            # Convert latent to CPU numpy for visualization
            latent_np = latent_state.detach().cpu().numpy()
            
            if latent_np.ndim == 4:  # [B, C, H, W]
                batch_size, channels, height, width = latent_np.shape
                
                # Channel selector
                imgui.text(f"Latent Shape: {latent_np.shape}")
                changed, selected_channel = imgui.slider_int(
                    "Channel", 0, 0, channels - 1
                )
                
                # Statistics
                channel_data = latent_np[0, selected_channel]  # First batch
                imgui.text(f"Min: {channel_data.min():.4f}")
                imgui.text(f"Max: {channel_data.max():.4f}")
                imgui.text(f"Mean: {channel_data.mean():.4f}")
                imgui.text(f"Std: {channel_data.std():.4f}")
                
                # Create heatmap visualization
                heatmap = create_latent_heatmap(channel_data)
                
                # Display in OpenCV window
                cv2.imshow("Latent Visualization", heatmap)
        
    imgui.end()
```

### Latent State Visualization
Create visual representations of internal agent states:

```python
def create_latent_heatmap(latent_channel, colormap=cv2.COLORMAP_VIRIDIS):
    """Create heatmap visualization of latent channel"""
    # Normalize to 0-255 range
    normalized = ((latent_channel - latent_channel.min()) / 
                  (latent_channel.max() - latent_channel.min() + 1e-8) * 255).astype(np.uint8)
    
    # Resize for better visibility
    resized = cv2.resize(normalized, (256, 256), interpolation=cv2.INTER_NEAREST)
    
    # Apply colormap
    heatmap = cv2.applyColorMap(resized, colormap)
    
    return heatmap

def visualize_memory_patterns(complex_memory):
    """Visualize complex-valued memory patterns"""
    if complex_memory is not None:
        memory_np = complex_memory.detach().cpu().numpy()
        
        # Separate real and imaginary parts
        real_part = np.real(memory_np)
        imag_part = np.imag(memory_np)
        
        # Create magnitude and phase visualizations
        magnitude = np.abs(memory_np)
        phase = np.angle(memory_np)
        
        # Normalize for display
        mag_vis = ((magnitude / magnitude.max()) * 255).astype(np.uint8)
        phase_vis = ((phase + np.pi) / (2 * np.pi) * 255).astype(np.uint8)
        
        # Apply colormaps
        mag_colored = cv2.applyColorMap(mag_vis, cv2.COLORMAP_HOT)
        phase_colored = cv2.applyColorMap(phase_vis, cv2.COLORMAP_HSV)
        
        # Combine side by side
        combined = np.hstack([mag_colored, phase_colored])
        
        return combined
    return None

def plot_attention_weights(attention_weights):
    """Visualize attention patterns from complex memory"""
    if attention_weights is not None:
        weights_np = attention_weights.detach().cpu().numpy()
        
        # Create bar chart visualization
        fig, ax = plt.subplots(figsize=(8, 4))
        ax.bar(range(len(weights_np)), weights_np)
        ax.set_xlabel('Memory Index')
        ax.set_ylabel('Attention Weight')
        ax.set_title('Memory Attention Pattern')
        
        # Convert matplotlib figure to OpenCV image
        fig.canvas.draw()
        img_array = np.frombuffer(fig.canvas.tostring_rgb(), dtype=np.uint8)
        img_array = img_array.reshape(fig.canvas.get_width_height()[::-1] + (3,))
        img_bgr = cv2.cvtColor(img_array, cv2.COLOR_RGB2BGR)
        
        plt.close(fig)
        return img_bgr
    return None
```

### Main Event Loop Integration
Integrate all interface components in the main simulation loop:

```python
def main_simulation_loop():
    """Main loop with integrated interface"""
    # Initialize components
    interface = RealTimeInterface()
    env = PyBulletEnvironment(use_gui=False)  # Use our custom interface instead
    agent = initialize_agent()
    optimizer = torch.optim.Adam(agent.parameters(), lr=1e-3)
    
    # Performance tracking
    training_metrics = {
        'loss_history': [],
        'fps': 0.0,
        'frame_count': 0,
        'start_time': time.time()
    }
    
    running = True
    while running:
        frame_start = time.time()
        
        # 1. Simulation step
        observation = env.get_observation()
        action = agent.forward(observation)
        next_observation, reward, done, info = env.step(action)
        
        # 2. Training step (if enabled)
        if interface.training_enabled:
            loss = agent.compute_loss(observation, action, reward, next_observation)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            
            training_metrics['loss_history'].append(loss.item())
        
        # 3. Render PyBullet camera
        simulation_frame = render_pybullet_camera(env.view_matrix, env.proj_matrix)
        
        # 4. Render ImGui interface
        interface.render_control_panel(agent, optimizer, env.constraint_manager)
        interface.render_performance_monitor(training_metrics)
        interface.render_latent_visualization(agent.current_latent)
        
        # 5. Display frames
        cv2.imshow("Simulation View", simulation_frame)
        
        # 6. Process events
        key = cv2.waitKey(1) & 0xFF
        if key == ord('q'):
            running = False
        elif key == ord(' '):  # Space to pause/unpause training
            interface.training_enabled = not interface.training_enabled
        elif key == ord('r'):  # Reset environment
            env.reset()
        
        # 7. Update performance metrics
        frame_time = time.time() - frame_start
        training_metrics['fps'] = 1.0 / frame_time if frame_time > 0 else 0.0
        training_metrics['frame_count'] += 1
        
        # 8. Render ImGui
        imgui.render()
        interface.impl.render(imgui.get_draw_data())
        
        # Episode reset handling
        if done:
            observation = env.reset()
    
    # Cleanup
    cv2.destroyAllWindows()
    p.disconnect()
```

## Acceptance Criteria
- [ ] OpenCV windows display PyBullet camera feed in real-time
- [ ] ImGui interface provides responsive parameter controls
- [ ] Training can be paused/resumed via interface
- [ ] Learning rate adjustments immediately affect optimizer
- [ ] Latent state visualizations update each frame
- [ ] Performance metrics (FPS, memory usage) displayed accurately
- [ ] Interface maintains 30+ FPS during training
- [ ] All controls preserve agent state when adjusted
- [ ] Clean shutdown and resource cleanup on exit

### Multi-Window Layout Manager
```python
class LayoutManager:
    def __init__(self):
        self.layouts = {
            'training': [('Control Panel', (0, 0, 300, 400)), 
                        ('Performance', (300, 0, 300, 200))],
            'debugging': [('Latent Viz', (0, 0, 400, 400)),
                         ('Memory Patterns', (400, 0, 400, 400))]
        }
        self.current_layout = 'training'
    
    def switch_layout(self, layout_name):
        if layout_name in self.layouts:
            self.current_layout = layout_name
```

### Custom Plotting Integration
```python
def render_custom_plots(training_history):
    """Render custom plots within ImGui"""
    if imgui.begin("Training Plots"):
        # Loss curves
        if len(training_history['jepa_loss']) > 0:
            imgui.plot_lines("JEPA Loss", training_history['jepa_loss'][-200:])
        
        if len(training_history['dsl_loss']) > 0:
            imgui.plot_lines("DSL Loss", training_history['dsl_loss'][-200:])
        
        # Constraint violations
        if len(training_history['constraint_violations']) > 0:
            imgui.plot_lines("Constraint Violations", 
                           training_history['constraint_violations'][-200:])
    
    imgui.end()
```

## Integration Notes
The interface will be integrated as the main visualization system:

```python
# Initialize interface instead of PyBullet GUI
interface = RealTimeInterface()
env = PyBulletEnvironment(use_gui=False)  # Headless PyBullet

# Main loop handles both simulation and visualization
main_simulation_loop()
```

This real-time interface provides **comprehensive monitoring and control** of the active inference agent, enabling live parameter adjustment, detailed internal state visualization, and performance monitoring during training and execution.
